keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,445 @@
+response is small (it fits a single page, assuming you won't make it extremely
+small), whereas the number of objects to process is proportional to cluster-size
+(so fairly big). For example, when kubelets requests pods schedule against it in
+a 5k node cluster with 30pods/node, the kube-apiserver must list the 150k pods
+from etcd and then filter that list down to the list of 30 pods that the kubelet
+actually need. This must occur for each list request from each of the 5k
+kubelets. If served from watch cache, this same request can be served by simply
+filtering out the 30 pods each kubelet needs from the data in the cache.
+
+In addition to the improvements to scale and performance, we aim to resolve a
+specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) remains open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted and this issue
+would be solved. "Going back in time" can curently happen if the initial list
+request is served from a stale watch cache with data much older than the
+reflector has previously observed or if the api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+## Proposal
+
+### Leveraging the Progress Notify Mechanism
+
+Guard this by a `WatchCacheConsistentReads` feature gate.
+
+#### Use WithProgressNotify to enable automatic watch updates
+
+Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+
+When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
+events to the watch automatically. By default etcd sends progress events every
+10 minutes, which is not frequent enough to be useful for our needs, so we will
+modify etcd to send them more frequently.
+
+When an consistent LIST request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
```

Comment:
This seems like a cascading failure waiting to happen :(  I think this fallback step could be rephrased as "If the cluster appears to be struggling, then do a more expensive operation."

If it's ok to do this fallback, then I counter-propose we just skip the cache for _all_ consistent reads.  At least in this case the apiserver/etcd cluster will be sized correctly, since the read load will be continuously apparent.

More constructively ;) I suggest we change this fallback step to "If the block times out, return GOAWAY", with a big comment explaining why we don't fallback to reading directly. Or alternatively (better?), just make `waitUntilFreshAndBlock` block forever (ie: until the client's own timeout gives up and retries/aborts).

Reply:
I don't think "goaway" is good enough proposal - I think that what you're just saying is that we need additional layer of rate-limitting at this point. Which is a fair point.

Reply:
You're right, we don't want to fallback to etcd. I've updated the doc and added a section mentioning both "Retry-After" and priority and fairness as options for dealing with rate-limiting clients.

Reply:
Revisiting this thread in light of https://github.com/kubernetes/kubernetes/pull/126139 / 

Struggling etcd *could* be *one* reason the block timed out, but it is not the only reason. *Anything* that makes the watch cache laggy will cause this failure mode, and there are known causes apart from struggling etcd that trigger that.

As a first step to ensure this feature doesn't break existing working users in new ways, @deads2k and I agree that we *should* fall back to serving from etcd (the existing code path) if the list-from-cache times out. If we instrument when that occurs, and instrument when the fallback succeeds, we'll get visibility to users and clusters that would be broken by defaulting this on with no fallback.